rfd/0159-password-status.md

```diff
@@ -0,0 +1,125 @@
+---
+authors: Bartosz Leper (bartosz.leper@goteleport.com)
+state: draft
+---
+
+# RFD 0159 - Password Status
+
+## Required Approvers
+
+- Engineering: (@zmb3 || @codingllama)
+- Security: @jentfoo
+- Product: @roraback
+
+## What
+
+We want to improve the UX of our flows related to changing passwords and management of authentication devices (both MFA and passwordless). The following new features must be supported:
+
+- The user must be able to set password having only passwordless credentials configured.
+- We should do our best to indicate which authentication methods a user has configured. Unfortunately, due to past design decisions, we will not be able to do it in every case; see the [Details](#details) section.
+
+## Why
+
+We are promoting passwordless authentication and would like to make it a default in our cloud environments. We need to decrease the UX friction and fill the missing gaps.
+
+Related issues:
+
+- [Design for defaulting to "Passwordless" authentication + updating view to clarify options](https://github.com/gravitational/cloud/issues/5757)
+- [Remove local password authentication from teleport](https://github.com/gravitational/teleport/issues/19671)
+
+## Details
+
+For the purpose of this document, we define "passwordless key/token" as a resident WebAuthn key.
+
+Currently, it's impossible to tell whether a user has a password configured. This is because the user always has _some_ password set. If they never set it up themselves, we just use a random 16-character password. As a consequence, when the user wants to set a password, we always ask them for the current one. It makes it impossible to add a password to account that already has a passwordless key configured and doesn't have any current password.
+
+### Changes to the Password Change Flow
+
+When the user wants to change their password, we need to first confirm their identity. Currently, we confirm the user's identity both by asking for the previous password, as well as performing a passkey/MFA token confirmation. To support adding a password where no password is configured, we propose to relax the confirmation conditions and allow setting password with _only_ passwordless token if such one has been used for confirming user's identity.
```

Comment:
> we propose to relax the confirmation conditions and allow setting password with _only_ passwordless token if such one has been used (...)

There's a wrinkle here that is worth discussing. If we are basing this decision on a resident credential, then the resulting WebAuthn assertion needs to have [UV=1][1] (User Verified) to be accepted. The surest way to guarantee that is to issue a challenge that requires User Verification ([userVerification=required][2]).

The wrinkle is that an userVerification=required challenge is different from a regular "MFA" challenge, which means that we must base challenge creation on the user's password state. The risk of using a regular "MFA" challenge is that the authenticator may choose to not do user verification, giving us a response with UV=0 even if the credential is resident (we could test various authenticators, I don't recall how they behave in this case). Also, according to the spec, [we should not fail operations with UV=0 if we didn't require it][3] in the first place.

Accepting a credential with UV=0 essentially means we are down a factor, so that's a no-go.

Please update the design to make sure we mention the appropriate UV requirements and how we handle those situations.

[1]: https://w3c.github.io/webauthn/#authdata-flags-uv
[2]: https://w3c.github.io/webauthn/#dom-authenticatorselectioncriteria-userverification
[3]: https://w3c.github.io/webauthn/#enumdef-userverificationrequirement

Reply:
@codingllama Thank you for pointing it out, that's indeed an important concern.

@roraback Since we have to decide if we request user verification prior to requesting authentication from the authenticator, I think we will have to explicitly distinguish between passwordless and MFA verification in the identity verification dialog. I'll tag you and Rishi in Figma and let's discuss the details there, where the UI context is more obvious.

Reply:
Something I was thinking about, but is admittedly a half-baked idea at this point, is that we could issue both kinds of challenges at once prior to a password change (only issuing the UV=true challenge if the user has capable devices). Maybe this could help with UI logic, although I still suspect you can only assert one of them at a time.

Reply:
@codingllama My idea was to alter the UX and make the user choose explicitly between the passwordless and MFA confirmation. But issuing both challenges at once? I don't see any place in the specs or other documentation that would convince me that this behavior is specified, and I'd rather avoid relying on unspecified behavior, even if it matches our expectations (which I doubt, I can bet we'd end up simply doing 2 ceremonies, or one of them if we decide to cancel the another one through a signal).

If anything, I'd rather think whether we are positively sure that `userVerification=preferred` (default) isn't actually the way to go. The [spec](https://w3c.github.io/webauthn/#dom-userverificationrequirement-preferred) says it means that "the [Relying Party](https://w3c.github.io/webauthn/#relying-party) prefers [user verification](https://w3c.github.io/webauthn/#user-verification) for the operation if possible". I guess our dilemma boils down to how exactly browsers interpret that "if possible" requirement. Unfortunately, this also puts us in a realm of not-entirely-specified behavior. :/

Reply:
Re challenges: just to be clear, I'm not suggesting that we lean into any unspecified behavior. Like you said, having the user choose the flow is probably the better solution.

Re userVerification=preferred: "preferred" means UV is a nice to have, but RP won't fail the operation if it's missing. In our case that would mean letting users change their passwords with a simple Yubikey tap, no PIN or biometrics required. I don't think we should let that happen without UV=1.

> I guess our dilemma boils down to how exactly browsers interpret that "if possible" requirement

The interpretation of the UV requirement is up to the authenticator. Some authenticators tend to always do UV (think biometrics), while PIN-based authenticators tend to skip the need for PINs unless explicitly required. I wouldn't count on them always behaving the way we expect with the "preferred" setting.

Reply:
@codingllama I think I found a solution that was right in front of us, and we just didn't see it. Here's how I see it.

For every user with at least 1 token, we can have the following scenarios:
1. The user has at least 1 MFA (non-resident) token. This means they _have to_ have a password, since it's not possible to enroll with a non-resident token and without password. It's also not possible to remove the existing password. (*) In this scenario, we can safely assume we can ask for the password, so we get the WebAuthn credentials, but don't enforce verification. From here, we can have two outcomes:
   1.1. The authenticator returned UV=1. We don't ask for the previous password, since the user has been verified. The backend sees UV=1 and lets us change password without the previous one.
   1.2. The authenticator returned UV=0. We ask for the previous password, since we're sure it exists. The backend sees UV=0 and ensures that the existing password is valid. Note that while the password change may not take an effect because of password validation, we still comply to the specs, because we don't fail the MFA ceremony itself.
2. The user has no MFA (non-resident) tokens, or in another words, all of their tokens are resident. This means the user may or may not have a password, so we can't require it. In this case, we get the WebAuthn credentials and enforce verification. We then ask for new password without confirming the old one. The backend sees UV=1 and lets us do it.

Note that in none of these scenarios, we don't risk locking the user out if they don't have physical access to some of their authenticators at a given moment. Case 2 still remains case 2; if the user can't access any device, they wouldn't be able to change the password anyway. If there is at least one device available, it still can provide user verification.

In scenario 1, if the user doesn't have access to any MFA, they either fall into a "I have no tokens" case, explained previously. If they have access to a passwordless token, they seem to fall into case 2, but at the same time, we know that the user has a password. If they don't know the password — currently they wouldn't be able to change the password, either. However, this shouldn't prevent us from making the change described by the RFD, as the goal of the change was not to allow setting a password without knowing previous one — the goal is to allow setting a password if none has been set before, and in scenario 1, we have proven that this is not the case. (Allowing users to set a password without giving the previous one is just means of achieving the goal, not the goal itself.)

One thing that I'm worried about (cc @roraback, @rishibarbhaya-design) is that when we use this logic for deciding whether to ask user for original password or not, we will end up confusing them and perhaps eroding trust. I think that on the screen where the user gives the old and new passwords, if we don't have the text field for the old password, we should place a note instead that says that we don't require it, because they strongly validated their identity. What do you think? (I have a meeting with Rishi today, we will discuss it 1:1.)

(*) The only edge case here is a user who is in the process of password reset, but this user won't be interacting with the account settings page until the procedure is done.

Reply:
(Actually, we could augment the password detection on login to include case 1 and simply use the password status flag on the account settings page to keep the system state a little bit more consistent.)

Reply:
> 1. The user has at least 1 MFA (non-resident) token. This means they have to have a password (...)

You definitely can enroll a non-resident token without a password. Just go the Account Settings page, for example. We can make this assumption with certainty.

> Note that in none of these scenarios, we don't risk locking the user out if they don't have physical access to some of their authenticators at a given moment

Lack of access to an authn factor is the purview of account recovery. We probably shouldn't mix those concerns with regular password changes (although passwordless account recovery is also in a sore spot).

I'm a bit concerned that we might be making this too evolved, although we could simplify the backend logic to something like this:

A user may change their password with:

1. A solved WebAuthn challenge with UV=1
2. A solved WebAuthn challenge with UV=0 plus their current password

How that translates in terms of UX I'm not sure.

Reply:
Correction, I meant this:

You definitely can enroll a non-resident token without a password. Just go the Account Settings page, for example. We __can't__ make this assumption with certainty.

Reply:
@codingllama OK, I stand corrected then. We'll need to make this change on the UX side.